SI-9776 Fix type of PriorityQueue.newBuilder and improve performance

chrisokasaki opened the pull request:
review by @Ichoran

Mainly fixes issues from SI-9776:
- Fixes the return type of mutable.PriorityQueue.newBuilder to be Builder[A, PriorityQueue[A]] rather than PriorityQueue[A].
- Improves performance of bulk inserts from O(N log N) to O(N), using the heapify algorithm instead of one-by-one inserts.  This primarily affects the builder, ++=, and reverse.  (These changes indirectly benefit the many, many other methods that use the builder or ++=.)

Also fixes issue from SI-9757:
- Fixes space leak in dequeue.  (However, there is still a similar space leak in clear.)

Non SI improvements:
- Slightly improves performance of clone.
- Improves scaladoc description of reverse, which was already out of date from an earlier commit.

Informal timings of the speedups using scalameter:
- Creating a priority queue of 1 million increasing numbers: speedup from 3098ms to 151ms!
- Creating a priority queue of 1 million random numbers: speedup from 338ms to 133ms.
- Reversing a priority queue of 1 million random numbers: speedup from 590ms to 219ms.

Ichoran commented on src/library/scala/collection/mutable/PriorityQueue.scala:
```diff
@@ -89,21 +89,23 @@ sealed class PriorityQueue[A](implicit val ord: Ordering[A])
         k = j
       }
     }
+    return k != m
```
This should just be `k != m` since it's the last expression.

Ichoran commented on src/library/scala/collection/mutable/PriorityQueue.scala:
```diff
@@ -142,9 +204,11 @@ sealed class PriorityQueue[A](implicit val ord: Ordering[A])
     if (resarr.p_size0 > 1) {
       resarr.p_size0 = resarr.p_size0 - 1
-      resarr.p_swap(1, resarr.p_size0)
+      val elementToReturn = resarr.p_array(1)
```
I think `result` is a more common name for temporary storage for a return value.

Ichoran commented on src/library/scala/collection/mutable/PriorityQueue.scala:
```diff
@@ -254,12 +320,24 @@ sealed class PriorityQueue[A](implicit val ord: Ordering[A])
-  override def clone(): PriorityQueue[A] = new PriorityQueue[A] ++= this.iterator
+  override def clone(): PriorityQueue[A] = {
+    val pq = new PriorityQueue[A]
+    for (i <- 1 to this.length) pq.unsafeAdd(resarr(i))
```
This is way less bad than the previous version, but why not just copy the underlying array?  For instance, dynamically expanding the array rather than just setting it to the correct size results in twice the allocation needed and 2.4ish the number of copies (depending on the distribution of the lengths).

Ichoran commented on src/library/scala/collection/mutable/PriorityQueue.scala:
```diff
@@ -254,12 +320,24 @@ sealed class PriorityQueue[A](implicit val ord: Ordering[A])
+    new Builder[A, PriorityQueue[A]] {
+      val pq = new PriorityQueue[A]
+      def +=(elem: A): this.type = { pq.unsafeAdd(elem); this }
+      def result(): PriorityQueue[A] = { pq.heapify(1); pq }
```
Can you decide whether this builder should be reusable or not (see the new ReusableBuilder trait)?  As written, the builder can be used once efficiently, but can still be added to and regenerated a little less efficiently (because heapify is called from 1, not only on the new elements).  Since most builder-use is single shot, this is probably the better way to go (as opposed to maintaining a list of number of elements stored since the last call to result).  In any case, this seems reusable, so maybe you can tag it as such.

chrisokasaki replied:
I think you're suggesting a behavior where you can add 1,2,3 to the builder, call result to get a priority queue with elements 1,2,3, then add 4,5,6 to the builder, and call result again to get a priority queue with elements 1,2,3,4,5,6.  But the priority queues are mutable, and I don't think you would want adding 4 to the builder to retroactively also add 4 to the first priority queue.  Or maybe I'm misunderstanding what you are saying.

Ichoran replied:
No, you understood me, I just wasn't saying anything sensible.  My mistake!  The builders for immutable collections are often reusable, but you're right, it doesn't make sense for this one to be.

Ichoran commented on src/library/scala/collection/mutable/PriorityQueue.scala:
```diff
@@ -186,25 +250,27 @@ sealed class PriorityQueue[A](implicit val ord: Ordering[A])
-   *  will be subsequent, but their corresponding subinterval may be inappropriately
-   *  reversed. However, due to the compare-equals contract, they will also be equal.
+   *  For example, the element with the highest priority in `pq` has the lowest
+   *  priority in `pq.reverse`, and vice versa.
```
I mostly prefer the new wording, but it doesn't say anything about what happens with ties.  Maybe just "Elements with equal priority will come off in arbitrary order (not necessarily the reverse of the original arbitrary order)," unless the new approach does guarantee exactly opposite order.  (I think not, in general?)

Ichoran commented on src/library/scala/collection/mutable/PriorityQueue.scala:
```diff
@@ -254,12 +320,24 @@ sealed class PriorityQueue[A](implicit val ord: Ordering[A])
+  override def clone(): PriorityQueue[A] = {
+    val pq = new PriorityQueue[A]
+    for (i <- 1 to this.length) pq.unsafeAdd(resarr(i))
+    pq
```
Maybe you want to mention, if you leave it this way, that you don't need to call `heapify` because the heap is already in order!  (Otherwise it's a bit jarring: "wait, didn't you say never to return after unsafeAdd without calling heapify?")

Ichoran commented on test/files/scalacheck/MutablePriorityQueue.scala:
```diff
@@ -0,0 +1,102 @@
+  property("clone") = forAll { list: List[E] =>
+    val pq = PriorityQueue(list : _*)
+    val c = pq.clone()
+    pq != c &&
```
Do you mean `(pq ne c)` here?

Ichoran commented:
Looks good; just a few minor comments.  Thanks!  This should speed PriorityQueue up a lot.

chrisokasaki commented:
I made all of those changes (except the reusable builder).  The speedup to clone was huge, nearly a factor of 10!  I was also able to use some of those same ideas in reverse, but there the improvement is much more modest because it still needs to heapify.

Ichoran commented:
Glad that made a big difference!  There's still the `pq != c` vs. `pq ne c` issue in the test, though (my last comment).  You want reference equality (the latter), right?

chrisokasaki commented:
Sigh.  Yes, I made the change but then accidentally left the test file out of the commit.

Ichoran commented on test/files/scalacheck/MutablePriorityQueue.scala:
```diff
@@ -0,0 +1,102 @@
+import Prop._
+import Arbitrary._
+
+object PQTest extends Properties("PriorityQueue") {
```
I think this has to be `object Test` to get run properly.  I'm not very familiar with ScalaCheck, but the other tests invariably use just `object Test` as the name.

Ichoran commented:
I think ScalaCheck isn't running the test because it wants `object Test` (not `PQTest`).  Also, if you could squash all commits (no value in keeping them separate; the history is cleaner with a single commit--though of course you can make your own branch that preserves them), the PR should be set to merge.

chrisokasaki commented:
Thanks, I couldn't make sense of the error reports.  I'll change the name and squash the commits.

chrisokasaki commented:
FYI, changing the name (PQTest to Test) and squashing the commits in one go is giving me conflicts, so I'm going try doing that in separate steps, which means that there's going to be at least one meaningless intermediate update before I can get it right.  Just ignore it. :-)

chrisokasaki commented:
Ok, here's the deal.  This is the first time I've ever used git (I normally use hg).  And like every git newbie, not only have I dorked things up far beyond my ability to fix them, but in my ignorant flailing around trying to fix things, I've also put things into an inconsistent state where these 6 commits CAN'T be squashed (or at least not by me).

The good news is that it's only 2 files.

From my end, the easiest thing to do would be to kill this pull request, and make a new one from scratch.  The main drawback is that it would lose the record of the conversation above.

Unless you have a better suggestion, that's my leaning.

retronym commented:
@chrisokasaki Here's the recipe to update this pull request with a single commit: https://github.com/scala/scala/pull/5184#issuecomment-220868526

chrisokasaki committed:
SI-9776: Fix type of PriorityQueue.newBuilder and improve performance
- Fix the return type of mutable.PriorityQueue.newBuilder to be
Builder[A, PriorityQueue[A]] rather than PriorityQueue[A].

- Improve performance of bulk inserts from O(N log N) to O(N),
primarily in the builder, ++=, and reverse.  These changes
indirectly benefit the many other methods that use the builder or ++=.

- Improve performance of clone.

- Fix SI-9757 space leak in dequeue.

chrisokasaki commented:
Awesome, that worked!  Thanks!

Ichoran commented:
LGTM

chrisokasaki commented:
Thanks for the help @Ichoran
